Internals: Improve coverage flow

gezalore opened the pull request:
Please see the added internals doc section on usage, but the high-level usage is:

```
./configure --enable-coverage
.. run whatever you want ..
make coverage-open
```

`make test` takes about 70 minutes on my system (~3x normal clean cache `make test`).
`make coverage-combine` (running fastcov) is ~90 seconds
`make coverage-report` (running genhtml) is ~30 seconds

You can see a codecov.io report here:
- https://app.codecov.io/gh/gezalore/verilator/tree/cover

Highlights:
- We no longer replace the opt build with the dbg build, we just build both with -O0 instead, compile time is not a big issue here, and end-to-end this is marginally faster than using the debug build all the time
- We disable optimization of the runtime lib as well, which we didn't before, coverage now looks much better
- Unused inline functions in headers now show as holes (most of the time)
- We now filter exclusions using fastcov, rather than the previous dance with a cloned source tree. This works well.
- We now upload the exclusion filtered report to codedcov.io (as opposed to the unfiltered one which we did before)
- Uploading to codecov.io no longer runs gcov agan on everything (--plugin noop), but just uses the lcov info file from fastcov (as was expected)
- Full dependency tracking of everything involved in coverage. If you run more test, just run `make coverage*` for the lastest
- CI now uses the reusable workflows for build and test
- Overall CI CPU time is ~20% less, latency is a few minutes shorter.

gezalore commented on configure.ac:
```diff
@@ -411,11 +412,19 @@ if test "$CFG_ENABLE_COVERAGE" = "yes"; then
+  _MY_CXX_CHECK_OPT(CXX,-fprofile-update=atomic)
+  # Ensure data files can be written from parallel runs
+  _MY_CXX_CHECK_OPT(CXX,-fprofile-reproducible=parallel-runs)
+  # Save source files as absolute paths in gcno files
```
FWIW these 2 options are what allows us to keep it simple. It enables updating the .gcda files from multiple invocations, and from multiple threads safely.

wsnyder commented on docs/guide/exe_verilator.rst:
```diff
@@ -780,7 +780,7 @@ Summary:
 
-   Feature may be one of the following: COROUTINES, SYSTEMC, ASAN.
+   Feature may be one of the following: COROUTINES, SYSTEMC, ASAN, GCOV.
 
```
GCOV feels wrong, we might have some other program do coverage or e.g. do MSVC coverage and should be same indication.

Also calling it COVERAGE (or even GCOV) can be confused with using coverage in SystemVerilog.  Maybe SELF-COVERAGE or something (suggest better naming).  Probably should use same --enable-self-coverage wording for config.

Nit: Let's sort this list.

gezalore replied:
Agree, I don't really like this, but I also don't like that developer only stuff is exposed in the docs here. I picked "GCOV" because for example `clang` has both a `gcov` based coverage flow, and a "Source-based Code Coverage" flow of its own (which seems superior, but didn't want to migrate to that right now), so might need to know the difference.

How about `DEV_GCOV` (and also `DEV_ASAN`) and then not include these in the user docs? One should never ship an asan or gcov build.

Then rename to `--enable-dev-gcov` and `--enable-dev-asan`? and change the configure macros `CFG_WITH_DEV_GCOV`? Not sure we want to touch `--enable-ccwarn`

gezalore replied:
Ok, I will do a pre-pull to rename the asan ones.

wsnyder commented on docs/internals.rst:
```diff
@@ -1656,6 +1656,40 @@ significant variance. Experience shows that a ~20% time difference can be
+   make coverage-open     # Create and open HTML coverate reprot
+
+
+The ``coverage-open`` make target depends on ``covarage-report``, which
+generates the HTML coverage report, and does in turn depend on
+``coverage-combine``, which combines all ``.gcda`` files into an lcov
+data file. You can use each of these make targets separately if you wish.
```
Coverage-open suggests it's starting coverage not viewing it.

Awkward, maybe:  (Avoid "you" - don't know who reader is.)

```suggestion
The ``coverage-view`` make target opens the generated coverage report. This depends on ``coverage-report``, which
generates the HTML coverage report. That in turn depends on
``coverage-combine``, which combines all ``.gcda`` files into an lcov
data file. Each of these make targets can be used separately if desired.
```

gezalore replied:
> Awkward

I know, I am, sorry ;) Fixed.

wsnyder commented on docs/internals.rst:
```diff
@@ -1656,6 +1656,40 @@ significant variance. Experience shows that a ~20% time difference can be
+Note that every time you invoke a binary built with coverage collection, it
+will add the coverage data collected during that invocation to the existing
+``.gcda`` files. This means that you can also just run Verilator on any other
+external files or projects to collect code coverage on those invocations.
+
```
Be aware if changing a test, if that test no longer covers some item the report will still contain the old coverage.  Use a make `coverage-zero` and rerun all tests if this is a concern.

wsnyder commented on include/verilated.mk.in:
```diff
@@ -22,6 +22,7 @@ PYTHON3 = @PYTHON3@
 # Configuration time options
 CFG_WITH_CCWARN = @CFG_WITH_CCWARN@
 CFG_WITH_LONGTESTS = @CFG_WITH_LONGTESTS@
+CFG_WITH_COVERAGE = @CFG_WITH_COVERAGE@
```
Sort, but might sort differently pending a rename of all this stuff.

```suggestion
CFG_WITH_CCWARN = @CFG_WITH_CCWARN@
CFG_WITH_COVERAGE = @CFG_WITH_COVERAGE@
CFG_WITH_LONGTESTS = @CFG_WITH_LONGTESTS@
```

wsnyder commented on nodist/code_coverage:
```diff
@@ -1,405 +0,0 @@
-                        lineno += 1
-                        line = line.rstrip()
-                        done = False
-                        if re.search(r'LCOV_EXCL_LINE', line):
-                            line += " LCOV_EXCL_BR_LINE"
-                            done = True
-                        elif re.search(r'LCOV_EXCL_START', line):
-                            line += " LCOV_EXCL_BR_START"
-                            done = True
-                        elif re.search(r'LCOV_EXCL_STOP', line):
-                            line += " LCOV_EXCL_BR_STOP"
-                            done = True
```
Does the toolage now handle this properly?  Did you compare all reports under old and new system to make sure all old disables like this, plus the ones in the .dat file were still properly suppressed?

wsnyder commented on .gitignore:
```diff
@@ -44,6 +44,7 @@ verilator_coverage_bin*
 /.vscode/
 /.idea/
 /cmake-build-*/
+/nodist/coverage/
```
obj_dir/coverage

or obj_coverage

Something that doesn't hide in what looks like source directory and `rm -rf obj*` will remove.

wsnyder commented on .github/workflows/build-test.yml:
```diff
@@ -36,6 +36,7 @@ jobs:
       os-name: linux
       cc: ${{ matrix.cc }}
       asan: ${{ matrix.asan }}
+      gcov: 0
```
```suggestion
      dev-gcov: 0
```

And dev-asan, so we are consistent.

wsnyder commented on docs/internals.rst:
```diff
@@ -1656,6 +1656,45 @@ significant variance. Experience shows that a ~20% time difference can be
+Code coverage
+-------------
+
+Code coverage for Verilator itself can be collected using ``gcc`` and ``gcov``
```
```suggestion
Code coverage for developing Verilator itself can be collected using ``gcc`` and ``gcov``
```

wsnyder reviewed (commented):
I prefer to keep fastcov as don't want to have a dependency, and also have tweaked it (and the upstream so it still matches upstream.)

Good idea, please add ci: dev-coverage,  could be separate pull if you prefer.

gezalore commented:
> I prefer to keep fastcov as don't want to have a dependency,

Uhm, it's kept (and I tweaked it further).

It's the `ci/codecov` binary that I deleted and replaced with https://github.com/codecov/codecov-action

gezalore commented:
> please add ci: dev-coverage,

the RTLMeter one is 'pr: rtlmeter' I picked 'pr: ' back then because the labels can be applied to issues as well, so wanted to highlight it should go on a PR. I'm happy to call it whatever you want, just highlighting the consistency issue.

wsnyder commented:
Oh, meant `pr:`
